### PR TITLE
Add integration tests for notifications

### PR DESCRIPTION
**Pull request by nickvergessen**
Signed-off-by: Joas Schilling <coding@schilljs.com>

### CODE REVIEW

**Review by danxuliu — COMMENTED** [review]
Rather than installing the notifications app only in Drone I would add:
```
NOTIFICATIONS_BRANCH="master"

# Add Notifications app to the parent directory of "spreed" (unless it is
# already there or in "apps").
if [ ! -e "${ROOT_DIR}/apps/notifications" ] && [ ! -e "../../../notifications" ]; then
       (cd ../../../ && git clone --depth 1 --branch ${NOTIFICATIONS_BRANCH} https://github.com/nextcloud/notifications)
fi
```
to the [`run.sh` file](https://github.com/nextcloud/spreed/blob/7a3a78b296f76a4e668097bcd9c32325988cfe17/tests/integration/run.sh#L21) to ensure that the notifications app is installed if needed, no matter if the integration tests are run in Drone, by directly calling `./run.sh` or in Docker with `./run-docker.sh`.

**Comment by nickvergessen** [comment]
We do it like this with guests app too atm

**Comment by nickvergessen** [comment]
Added installing of both apps in run.sh and added another check for additional app directories

File: tests/integration/spreedcheats/lib/Controller/ApiController.php [hunk 1]
```diff
@@ -85,6 +85,15 @@ public function resetSpreed(): DataResponse {
...
 			->executeStatement();
 
+		try {
+			$delete = $this->db->getQueryBuilder();
+			$delete->delete('notifications')
+				->where($delete->expr()->eq('app', $delete->createNamedParameter('spreed')))
+				->executeStatement();
+		} catch (\Throwable $e) {
+			// Ignore
+		}
```
Comment (danxuliu) [inline tests/integration/spreedcheats/lib/Controller/ApiController.php:95]: The other queries are not guarded against exceptions, should they be guarded too or they will always succeed and this one has something special that could cause the query to throw in some cases?

File: tests/integration/features/chat/notifications.feature [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+  Background:
+    Given user "participant1" exists
+    Given user "participant2" exists
+
+  Scenario: Normal message when recipient is online in the one-to-one
+    When user "participant1" creates room "one-to-one room" (v4)
+      | roomType | 1 |
+      | invite   | participant2 |
+    Given user "participant2" joins room "one-to-one room" with 200 (v4)
```
Comment (danxuliu) [inline tests/integration/features/chat/notifications.feature:11]: I would say that everything before sending the message should be a precondition, and therefore the steps should be prefixed by `Given` and `And` (here and in all the scenarios of this file), but I guess that is just a matter of taste :-)

**Review by danxuliu — APPROVED** [review]
Other than the comments, :+1: 

I have added a fixup to remove resetting the current user in _user :user has the following notifications_, as it is not needed and inconsistent with the rest of steps that get the user as a parameter, and another fixup to simplify cloning the needed apps, as it should not be needed to manually check the paths if `app:getpath` returned false.

File: tests/integration/spreedcheats/lib/Controller/ApiController.php (hunk 1 above)
Comment (nickvergessen) [reply tests/integration/spreedcheats/lib/Controller/ApiController.php:95]: This one is guarded as it is against the table of another app and I thought better save then annoyed

**Pull request merged**